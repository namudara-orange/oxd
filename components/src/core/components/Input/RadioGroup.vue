<script lang="ts">
import {defineComponent, h} from 'vue';
import RadioInput from '@orangehrm/oxd/core/components/Input/RadioInput.vue';
import InputGroup from '@orangehrm/oxd/core/components/InputField/InputGroup.vue';
import useTranslate from '../../../composables/useTranslate';

export interface State {
  focused: boolean;
}

export interface Options {
  id: number;
  label: string;
  disabled?: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  style?: Record<string, any>;
}

export default defineComponent({
  name: 'oxd-radio-group',
  components: {
    RadioInput,
    InputGroup,
  },

  emits: ['update:modelValue', 'blur', 'focus', 'change'],

  props: {
    id: {
      type: String,
      default: 'radio-group-id',
    },
    class: {
      type: String,
      default: 'radio-column',
    },
    style: {
      type: Object,
    },
    disabled: {
      type: Boolean,
    },
    options: {
      type: Array,
    },
    modelValue: {
      type: String,
    },
  },

  render() {
    const inputId = this.id == '' ? 'radio-group-id' : this.id;
    const inputClass = this.class == '' ? 'radio-column' : this.class;
    const {$t} = useTranslate();
    return h(
      InputGroup,
      {
        classes: {
          wrapper: inputClass,
        },
        id: inputId,
        style: this.style,
      },
      [
        this.options.map((option: Options, index: number) => {
          return h(RadioInput, {
            id: inputId + '_' + option.id,
            style: option.style,
            autofocus: this.$attrs.autofocus && index === 0 ? true : false,
            optionLabel: $t(option.label),
            value: option.id,
            modelValue: this.modelValue,
            disabled: this.disabled ? 'true' : option.disabled,
            onFocus: () => {
              this.$emit('focus', true);
            },
            onBlur: () => {
              this.$emit('blur', true);
            },
            onClick: () => {
              this.$emit('click', true);
            },
            onChange: () => {
              this.$emit('update:modelValue', option.id);
            },
          });
        }),
      ],
    );
  },
});
</script>

<style lang="scss" scoped>
:deep(.radio-row) {
  display: flex;
  gap: 2rem;
}

:deep(.radio-column) {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}
</style>
