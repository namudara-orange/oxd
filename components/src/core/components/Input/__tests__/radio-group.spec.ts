import {mount} from '@vue/test-utils';
import RadioGroup from '@orangehrm/oxd/core/components/Input/RadioGroup.vue';

const optionsList = [
  {
    id: 1,
    label: 'Item one',
  },
  {
    id: 2,
    label: 'Item two',
  },
];

describe('RadioGroup > RadioGroup.vue', () => {
  it('renders OXD RadioGroup', () => {
    const wrapper = mount(RadioGroup, {
      props: {
        options: optionsList,
        class: 'radio-column',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should accept valid input', async () => {
    const wrapper = mount(RadioGroup, {
      props: {
        options: optionsList,
      },
    });
    wrapper.find('#radio-group-id_1').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toEqual([[1]]);
  });

  it('should emit the new value on click another radio button', async () => {
    const wrapper = mount(RadioGroup, {
      props: {
        options: optionsList,
        modelValue: 1,
      },
    });
    wrapper.find('#radio-group-id_2').trigger('click');
    expect(wrapper.emitted('update:modelValue')).toEqual([[2]]);
  });

  it('should focus the checkbox', async () => {
    const wrapper = mount(RadioGroup, {
      props: {
        options: optionsList,
      },
    });
    wrapper.find('#radio-group-id_1').trigger('focus');
    expect(wrapper.emitted('focus')).toEqual([[true]]);
  });

  it('should blur the checkbox', async () => {
    const wrapper = mount(RadioGroup, {
      props: {
        options: optionsList,
      },
    });
    wrapper.find('#radio-group-id_1').trigger('blur');
    expect(wrapper.emitted('blur')).toEqual([[true]]);
  });

  it('should change the class and id', async () => {
    const wrapper = mount(RadioGroup, {
      props: {
        options: optionsList,
        id: 'radio-new-id',
        class: 'radio-new-class',
      },
    });

    expect(wrapper.find('.radio-new-class').exists()).toBeTruthy();
    expect(wrapper.find('#radio-new-id_1').exists()).toBeTruthy();
  });
});
