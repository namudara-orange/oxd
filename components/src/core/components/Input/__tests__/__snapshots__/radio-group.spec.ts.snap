// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup > RadioGroup.vue renders OXD RadioGroup 1`] = `
<div class="oxd-input-group">
  <div class="oxd-input-group__label-wrapper">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </div>
  <div class="radio-column">
    <div class="oxd-radio-wrapper"><label class="">
        <!--v-if--><input type="radio" id="radio-group-id_1" value="1"><span class="oxd-radio-input oxd-radio-input--active --label-right oxd-radio-input"></span>Item one
      </label></div>
    <div class="oxd-radio-wrapper"><label class="">
        <!--v-if--><input type="radio" id="radio-group-id_2" value="2"><span class="oxd-radio-input oxd-radio-input--active --label-right oxd-radio-input"></span>Item two
      </label></div>
  </div>
  <!--v-if-->
  <!--v-if-->
</div>
`;
